pkg/controller/clustercatalog_controller.go

```diff
@@ -0,0 +1,401 @@
+package controller
+
+import (
+	"fmt"
+
+	patch "github.com/evanphx/json-patch"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-olm-operator/pkg/clients"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/management"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+
+	catalogdv1alpha1 "github.com/operator-framework/catalogd/api/core/v1alpha1"
+)
+
+const (
+	reasonCreated    = "Created"
+	reasonPatched    = "Patched"
+	reasonAsExpected = "AsExpected"
+)
+
+func NewClusterCatalogController(name string, manifest []byte, operatorClient *clients.OperatorClient, dynamicClient dynamic.Interface, clusterCatalogClient *clients.ClusterCatalogClient, recorder events.Recorder) (factory.Controller, error) {
+	obj, _, err := scheme.Codecs.UniversalDecoder().Decode(manifest, nil, &unstructured.Unstructured{})
+	if err != nil {
+		return nil, fmt.Errorf("decoding manifest data: %w", err)
+	}
+
+	uObj := obj.(*unstructured.Unstructured)
```

Comment:
These code pattern looks different than the other controllers we have e.g. pkg/controller/incompatible_operator_controller.go or pkg/controller/staticupgradeableconditioncontroller.go . Just to make sure this is an explicit choice.

Reply:
This code is now completely different